---
title: "PEC1 Análisis de datos ómicos"
subtitle: 'Máster interuniversitario de Bioestadística y Bioinformática'  
author: "Luis Morís Fernández"
date: "`r Sys.Date()`"
output:
  tufte::tufte_html: default
  tufte::tufte_handout:
    citation_package: natbib
    latex_engine: xelatex
bibliography: skeleton.bib
link-citations: yes
---


```{r setup, include=FALSE}
library(tufte)
# invalidate cache when the tufte version changes
knitr::opts_chunk$set(tidy = FALSE, cache.extra = packageVersion('tufte'))
options(htmltools.dir.version = FALSE)
library(GEOquery)
library(Biobase)
library(arrayQualityMetrics)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(magick)
library(oligo)
library(limma)
```

```{r parameters_functions, echo=FALSE}
GEO_ID = 'GSE7614'
GEO_DOI =  '10.1016/j.neurobiolaging.2006.08.004'
PLATFORM_URL = 'https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL1322'
PAPER_URL = 'https://genomebiology.biomedcentral.com/articles/10.1186/gb-2007-8-10-r231'



plotPCA3 = function (datos, labels, factor, title, scale,colores, size = 1.5, glineas = 0.25) {
   data = prcomp(t(datos),scale=scale)
   # plot adjustments
   dataDf = data.frame(data$x)
   Group = factor
   loads = round(data$sdev^2/sum(data$sdev^2)*100,1)
   # main plot
   p1 = ggplot(dataDf,aes(x=PC1, y=PC2)) +
     theme_classic() +
     geom_hline(yintercept = 0, color = "gray70") +
     geom_vline(xintercept = 0, color = "gray70") +
     geom_point(aes(color = Group), alpha = 0.55, size = 3) +
     coord_cartesian(xlim = c(min(data$x[,1])-5,max(data$x[,1])+5)) +
     scale_fill_discrete(name = "Group")
   # avoiding labels superposition
   p1 + geom_text_repel(aes(y = PC2 + 0.25, label = labels),segment.size = 0.25, size = size) + 
     labs(x = c(paste("PC1",loads[1],"%")),y=c(paste("PC2",loads[2],"%"))) +  
     ggtitle(paste("Principal Component Analysis for: ",title,sep=" "))+ 
     theme(plot.title = element_text(hjust = 0.5)) +
     scale_color_manual(values=colores)
}


# A volcano plot with some more bells and whistles
my_volcano_plot = function(cm, geneSymbols){
  
  # cm the ouput of the ebayes function
  # genesymbols: select(drosophila2.db, rownames(cm), c("SYMBOL"))
  
  plot_logod = tibble::as_tibble(cm$coefficients, rownames='PROBEID') %>%
  tidyr::pivot_longer(-PROBEID, names_to = 'contrast', values_to = 'coef')

plot_d = tibble::as_tibble(cm$p.value, rownames='PROBEID') %>%
  tidyr::pivot_longer(-PROBEID, names_to = 'contrast', values_to = 'p_val') %>% 
  dplyr::group_by(contrast) %>% dplyr::mutate(adj_p_val = p.adjust(p_val, method = 'fdr')) %>% dplyr::ungroup() %>% 
  dplyr::mutate(sig_p = p_val<.05,
                sig_adj_p = adj_p_val<.05,
                minus_log_10_p = -log10(p_val),
                minus_log_10_adj_p = -log10(adj_p_val),) %>% 
  dplyr::inner_join(geneSymbols, by='PROBEID') %>% 
  dplyr::inner_join(plot_logod, by=c('PROBEID', 'contrast'))

unadj_p = ggplot(plot_d, aes(x=coef, y=minus_log_10_p, color=sig_p)) +
  geom_point() +
  ggthemes::theme_tufte(base_family = 'sans') +
  geom_hline(yintercept = -log10(.05), linetype=2) + 
  geom_vline(xintercept = 1) + 
  geom_vline(xintercept = -1) + 
  facet_wrap(.~contrast, scale='free') +
  labs(x = 'Coefficient', y = '-log10(p)', color = 'unadjusted p<.05', title='Unadjusted p-value')

adj_p = ggplot(plot_d, aes(x=coef, y=minus_log_10_adj_p, color=sig_adj_p)) +
  geom_point() +
  ggthemes::theme_tufte(base_family = 'sans') +
  geom_hline(yintercept = -log10(.05), linetype=2) + 
  geom_vline(xintercept = 1) + 
  geom_vline(xintercept = -1) + 
  facet_wrap(.~contrast, scale='free')+
  labs(x = 'Coefficient', y = '-log10(p_adj)', color = 'FDR Corrected p<.05', title='FDR Corrected p-value')

  return(list(unadj_plot = unadj_p, adj_plot=adj_p, plot_d=plot_d))
}


```


# Información sobre el dataset

Vamos a utilizar el dataset `r GEO_ID`. Este dataset corresponde al estudio *Phenotypic and transcriptional response to selection for alcohol sensitivity in Drosophila melanogaster*.

|Información del dataset   |
|-----+-------|
| Organismo | [Drosophila Melanogaster](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=7227)|
| Número de muestras | 24 (15 individuos por muestra)|
|Grupos| Macho Control (x2 Rep. Biológicas x2 Líneas)|
| | Hembra Control (x2 Rep. Biológicas x2 Líneas)|
| | Macho Sensible (x2 Rep. Biológicas x2 Líneas)|
| | Hembra Sensible (x2 Rep. Biológicas x2 Líneas)|
| | Macho Resistente (x2 Rep. Biológicas x2 Líneas)|
| | Hembra Resistente (x2 Rep. Biológicas x2 Líneas)|
|Número de características | 18952 |
| Plataforma| [Drosophila_2] Affymetrix Drosophila Genome 2.0 Array|
| |[GPL1322](`r PLATFORM_URL`)|
|Tipo de muestra| *in situ oligonucleotide* |


```{r readCELFiles, echo=FALSE, cache=TRUE, warning=FALSE, message=FALSE}

targets = readr::read_csv('data/targets.csv',col_types = 'cccdccc')

cel_files = list.celfiles('./data/', full.names=T)

my_targets = read.AnnotatedDataFrame('./data/targets.csv', header = T, row.names = 1, sep = ',')

sink('/dev/null') # Avoid output
raw_data = read.celfiles(cel_files, phenoData=my_targets, verbose = T)
sink()

# gse = getGEO(GEO_ID, GSEMatrix = T)[[1]]
# save(gse,file='dataset.RData')
# load('dataset.RData')
# gse
# glimpse(pData(gse))
# exprs(gse)

rownames(pData(raw_data)) = my_targets@data$short_name
colnames(raw_data) = rownames(pData(raw_data))
```

# Cargado de datos crudos

En este caso hemos descargado todos los archivos .CEL del dataset `r GEO_ID`. ^[[https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE7614](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE7614)]
También hemos consultado a que grupo pertenece cada una de las 24 muestras que contiene el dataset. Hemos preparado también un archivo que contiene toda la información referente a los grupos y las muestras y lo hemos cargado en memoria.

```{r show_targets, echo=FALSE}
head(targets) %>% kableExtra::kable(caption='Muestra del archivo que contiene *phenoData*')
```

Luego hemos generado un ExpressionFeatureSet juntando los archivos .CEL y annotatedDataFrame que contiene el *phenoData*
```{r show_raw_data, echo=FALSE}
suppressMessages(raw_data)
```

Vamos ahora a hacer un control de calidad de estos datos crudos.

# Control de calidad de los datos crudos

```{r compute_array_quality_metrics, echo=FALSE}
# Lo hacemos con el Array Quality Metrics, queda comentados para no ejecutarlos continuamente
# Luego sacaremos los gráficos de ahí
# arrayQualityMetrics::arrayQualityMetrics(raw_data, intgroup = c('sex', 'selection', 'line'), outdir = 'pre_norm')
# arrayQualityMetrics::arrayQualityMetrics(raw_data, intgroup = c('sex', 'selection', 'line'), outdir = 'pre_norm_log', do.logtransform = T)
# exprs(raw_data)
```

Para realizar un control de la calidad de los datos crudos vamos utilizar la libreria Array Quality Metrics y affyPLM.

Lo haremos de dos maneras diferentes, primero sin transformar los datos con logaritmo de la intensidad y luego haciéndolo.

## Sin transformar


```{r plot_raw_data, echo=FALSE, fig.fullwidth = TRUE, out.width='100%', fig.height=6, out.height='100%', fig.cap='Salida Array Quality Metrics sin transformar'}
out = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read("./pre_norm/outlier.png"))))
hm = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm/hm.pdf"))))
ma = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm/ma.pdf"))))
msd = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm/msd.pdf"))))
pca = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm/pca.pdf"))))
dens =(cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm/dens.pdf"))))
box = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm/box.pdf"))))
plot_grid(out, hm, box, pca, dens, msd, ma, ncol = 3)
``` 


- En el gráfico A podemos observar un resumen de todas las muestras y el resultado obtenido del control de outliers. Vemos que según el algoritmo las muestras 1 y 17 podrían presentar algún problema por ello las tendremos en especial consideración cuando observemos los gráficos. En este caso no disponemos de la posibilidad de hacerlo, pero podríamos comprobar las imágenes de los microarrays para ver si ha ocurrido alǵun problema durante su procesado.

- En el gráfico B observamos una agrupación por distancias de los microarrays, vemos que hay una separación clara según el sexo de la muestra, a excepción de las muestras 1 y 17 que muestran una distancia un poco atípica se encuentran más distantes del resto de muestras de machos.

- En el gráfico C observamos los boxplots de las intensidades de cada uno de los arrays. También aparece marcado el array quizá más disperso que el resto, también aparece marcado el 10, sin embargo no parece que sean muestras particularmente aberrantes.

- En el gráfico D podemos observar un análisis por componentes principales, veremos debajo uno como el que se ha realizado durante las prácticas de la asignatura, en este caso vemos una separación clara entre los grupos macho y hembra de las muestras lo cual es esperable y deseable.

- En el gráfico E, vemos que las muestra siguen un mismo patrón, aunque la escala hace un poco complicada esta observación, probaremos luego con una función de affysimple.

- En el gráfico F, el valor esperable es que la mediana sea constante, sin embargo aparece una pequeña joroba en el lado derecho, pero seguramente esté relacionado con una saturación de la intensidades como indica el propio paquete.

- En el gráfico G, podemos observar la comparación de cada uno de los arrays respecto a un pseudo-array que tiene el valor mediano a través del resto de arrays. En este caso lo esperable es que los valores se concentren a lo largo del eje M=0. Sin embargo al no estar transformados los valores aparece agrupados en el margen izquierdo haciendo difícil la visualización.

Vamos ahora a ver un PCA más completo

```{r PCA_raw, echo=FALSE, fig.margin=TRUE}
plotPCA3 (exprs(raw_data), labels=pData(raw_data)$short_name, factor=pData(raw_data)$selection, title='', scale=F,colores=c("red", "blue", "green", "yellow",'brown','pink'), size = 1.5, glineas = 0.25)
```

## Log-Transform

```{r plot_log_data, echo=FALSE, fig.fullwidth = TRUE, out.width='100%', fig.height=6, out.height='100%', fig.cap="Array Quality Metrics Output",fig.fullwidth = TRUE}
out = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read("./pre_norm_log/outlier.png"))))
hm = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm_log/hm.pdf"))))
ma = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm_log/ma.pdf"))))
msd = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm_log/msd.pdf"))))
pca = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm_log/pca.pdf"))))
dens =(cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm_log/dens.pdf"))))
box = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./pre_norm_log/box.pdf"))))
plot_grid(out, hm, box, pca, dens, msd, ma, labels = 'AUTO', label_size = 7)
``` 


Al estar transformados con el logaritmo la forma de las gráficas cambia ya que hemos *estirado* la distribución. El significado de los gráficos es el mismo que en el apartado anterior.

El array 17 sigue apareciendo como un outlier, sin embargo en ninguno de los plots aparece particularmente dramático por tanto lo conservaremos.

## affyPLM

Vamos ahora a utilizar la librería *affyPLM* vamos a ajusjar un modelo a nivel de sonda y vamos a comprobar las gráficas *RLE* y *NUSE*.

En el caso del *RLE* vamos a encontrar una medida estandarizada de la expresión, que ha de ser similar a través de arrays, y en el caso de *NUSE* observaremos los residuos normalizados del modelo cuya distribución ha de ser similar a través de arrays.

```{r show_affymetrics, include=FALSE,fig.fullwidth = TRUE}
library(affyPLM)

if(!file.exists('affyPLM.RData')){
  fit1 <- oligo::fitProbeLevelModel(raw_data)
  save(fit1, file='affyPLM.RData')
}else{load('affyPLM.RData')}

RLE_d = oligo::RLE(fit1) %>%
  tibble::as_tibble(RLE_d, rownames='probe') %>%
  tidyr::pivot_longer(-probe, names_to = 'short_name', values_to = 'intensity') %>%
  tidyr::separate(short_name, into=c('group', 'line', 'sex','id'), sep='_', remove=FALSE)

NUSE_d = oligo::NUSE(fit1) %>%
  tibble::as_tibble(RLE_d, rownames='probe') %>%
  tidyr::pivot_longer(-probe, names_to = 'short_name', values_to = 'intensity') %>%
  tidyr::separate(short_name, into=c('group', 'line', 'sex','id'), sep='_', remove=FALSE)

RLE_p = ggplot(RLE_d, aes(x=short_name, y=intensity, color=sex, fill=group)) +
  geom_boxplot(outlier.shape = NA) +
  scale_color_grey() +
  ggthemes::theme_tufte(base_family = 'sans') +
  theme(axis.text.x = element_text(angle = 90), legend.position = 'none') +
  scale_y_continuous(breaks = c(-0.5,0,0.5), limits = c(-.75, .75)) +
  labs(title = 'RLE', x = 'sample')

NUSE_p = ggplot(NUSE_d, aes(x=short_name, y=intensity, color=sex, fill=group)) +
  geom_boxplot(outlier.shape = NA) +
  scale_color_grey() +
  ggthemes::theme_tufte(base_family = 'sans') +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(limits = c(0.95, 1.05)) +
  labs(title = 'NUSE', x = 'sample')

plot_grid(RLE_p, NUSE_p)
```

No parece que ninguno de los arrays se desvíe particularmente en ninguna de las dos gráficas, si que vemos que hay una diferencia clara entre sexos. Podríamos preguntarnos en este caso si se ha cometido algún error y los machos perteneces a un batch y las hembras a otros, es decir hemos generado algún tipo de confound, aunque es probable que se deba a diferencias reales.

## Conclusiones de calidad

- Conservamos todos los arrays
- Podría ser útil revisar el 17
- Podría ser útil consultar si ha habido alguna diferencia sistemática entre los arrays de machos y hembras, batch, fecha, operario,...


# Normalización

Procedemos ahora a normalizar los datos con el método *RMA*.

Durante este paso el algoritmo:

- Eliminará el ruido de fondo
- Normalizará la señal entre arrays, nunca entre condiciones experimentales.
- Colapsará todos los valores de cada grupo de sondas en un único valor por gen.

```{r compute_rma, cache=T}

if(!file.exists('rma_normalized.Rda')){
  eset_rma = oligo::rma(raw_data)
  save(eset_rma, file="rma_normalized.Rda")
}else{
  load(file="rma_normalized.Rda")
}
```

```{r compute_AQM_rma}
# arrayQualityMetrics::arrayQualityMetrics(eset_rma, intgroup = c('sex', 'selection', 'line'), outdir = 'norm', force=T)
```

Una vez normalizado vamos a volver a realizar un análisis de calidad de los datos

## Array quality metrics output después de la normalización

```{r show_AQM_rma, echo=FALSE, fig.fullwidth = TRUE, out.width='100%', fig.height=6, out.height='100%', fig.cap="Array Quality Metrics Output",fig.fullwidth = TRUE}
out = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read("./norm/outlier.png"))))
hm = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./norm/hm.pdf"))))
ma = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./norm/ma.pdf"))))
msd = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./norm/msd.pdf"))))
pca = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./norm/pca.pdf"))))
dens =(cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./norm/dens.pdf"))))
box = (cowplot::ggdraw() + cowplot::draw_image(image_trim(image_read_pdf("./norm/box.pdf"))))
plot_grid(out, hm, box, pca, dens, msd, ma, labels = 'AUTO', label_size = 7)
```

Los gráficos son los mismos que en apartados anteriores, ahora vemos una separación mucho más clara en el array de distancias. El array número 17 ahora es más similar al resto.

Solo existen dos detalles destacables, uno es la diferencia entre patrones de intensidad en la figura E entre las muestras machos y hembras, y la joroba que aparece en la figura F, cuando la mediana debería de ser plana, quizá sea debida a la diferencia entre sexos.


## Conclusiones normalización

Los datos parecen correctos, por tanto no es necesaria ninguna accción.


# Filtrado

Vamos a filtrar las probes que no tengan entrada en y las que corresponden a las sondas de control de Affymetrix. Vamos a hacerlo de manera manual utilizando la base de datos correspondiente al chip: *Affymetrix Drosophila Genome 2.0 Array*, en el paquete 'drosophila2.db'. Lo que haremos será eliminar todas las *features* cuyo entrada *EntrezID* sea *NA*. 

```{r filtrado no específico, echo = FALSE}
library(drosophila2.db)
geneSymbols <- suppressMessages(select(drosophila2.db, rownames(eset_rma), c("SYMBOL", "ENTREZID", "GENENAME")))
filtered_probes = geneSymbols %>% dplyr::filter(!is.na(ENTREZID), !is.na(GENENAME)) %>% dplyr::pull(PROBEID)

eset_rma_filtered = eset_rma[rownames(eset_rma) %in% filtered_probes, ]
```

De esta manera pasamos de tener `r nrow(eset_rma)` *features* a `r nrow(eset_rma_filtered)`. Lo cual nos ayudará, por ejemplo, reduciendo el impacto de la correción por múltiples comparaciones.

# Comparaciones

Estamos ante un diseño de 3x2, en este caso vamos a simular que estamos interesados en las comparaciones:

- Main Effect Macho vs. Hembra
- Single Effect Resistente vs Sensible
- Interacción Sexo vs (Resistente vs Sensible)

```{r compute_comparions, echo=FALSE}
design_matrix = model.matrix(~0+group, data=pData(eset_rma_filtered))
colnames(design_matrix)=c('control_female','control_male','resistant_female','resistant_male','sensitive_female','sensitive_male')

contrast_matrix = makeContrasts(res_vs_sens = resistant_female + resistant_male - sensitive_female - sensitive_male, #Regardless of sex
                                male_vs_female = resistant_female + control_female + sensitive_female -
                                                 (resistant_male + control_male + sensitive_male), # ME of sex
                                int_sex_resvssens = (resistant_female - resistant_male) - (sensitive_female - sensitive_male),
                                levels = design_matrix)

m = lmFit(eset_rma_filtered, design_matrix)
cm = eBayes(contrasts.fit(m, contrast_matrix))
top_sex = topTable(cm, coef='male_vs_female', adjust='fdr', number=10000, sort.by = 'p')
top_res_vs_ses = topTable(cm, coef='res_vs_sens', adjust='fdr', number=10000, sort.by = 'p')
top_int = topTable(cm, coef='int_sex_resvssens', adjust='fdr', number=10000, sort.by = 'p')
```

Vamos a ajustar un modelo lineal con Limma y la siguiente matriz de diseño

```{r show_design_matrix, fig.cap='Horizontal axis indicates the term of the design matrix. Vertical axis indicate the short name of the sample'}

heatmap(design_matrix, Colv = NA, Rowv=NA,main='Contrast Matrix', col=c('black', 'white'))

# design_matrix_tib = tibble::as_tibble(design_matrix, rownames = 'Sample') %>%
#   dplyr::mutate(control_female = kableExtra::cell_spec(control_female, 'html', bold = control_female==1),
#                 control_male = kableExtra::cell_spec(control_male, 'html', bold = control_male==1),
#                 resistant_female = kableExtra::cell_spec(resistant_female, 'html', bold = resistant_female==1),
#                 resistant_male = kableExtra::cell_spec(resistant_male, 'html', bold = resistant_male==1),
#                 sensitive_female = kableExtra::cell_spec(sensitive_female, 'html', bold = sensitive_female==1),
#                 sensitive_male = kableExtra::cell_spec(sensitive_male, 'html', bold = sensitive_male==1),
#                 )
# design_matrix_tib %>% kableExtra::kable(format='html', escape=F, caption='Design Matrix') %>% kableExtra::kable_styling(full_width = F) %>% kableExtra::row_spec(0, angle=-45) %>% kableExtra::add_header_above(c(' '=1, 'Model Term'=6))


```

Una vez ajustado el modelo calculamos los siguientes contrastes

```{r show_contrast_matrix}
# contrast_matrix %>% kableExtra::kable(caption='Contrast Matrix') %>% kableExtra::kable_styling()

heatmap(t(contrast_matrix), Colv = NA, Rowv=NA,main='Contrast Matrix', col=c('#FFFFFF', '#666666', '#000000'))
legend(x="topleft", legend=c("1", "0", "-1"),
     fill=c('#FFFFFF', '#666666', '#000000'))
```


# Anotación

Anotamos los datos con el paquete *drosophila2.db*.

```{r annotation, echo=FALSE}
library(drosophila2.db)
annotatedTopTable <- function(topTab, anotPackage)
{
  topTab <- cbind(PROBEID=rownames(topTab), topTab)
  myProbes <- rownames(topTab)
  thePackage <- eval(parse(text = anotPackage))
  geneAnots <- suppressMessages(select(thePackage, myProbes, c("SYMBOL", "ENTREZID", "GENENAME")))
  annotatedTopTab<- merge(x=geneAnots, y=topTab, by.x="PROBEID", by.y="PROBEID")
return(annotatedTopTab)
}

top_sex_annotated = annotatedTopTable(top_sex, anotPackage = 'drosophila2.db') %>% dplyr::arrange(P.Value)
top_int_annotated = annotatedTopTable(top_int, anotPackage = 'drosophila2.db') %>% dplyr::arrange(P.Value)
top_res_vs_sens_annotated = annotatedTopTable(top_res_vs_ses, anotPackage = 'drosophila2.db') %>% dplyr::arrange(P.Value)


```


# Resultados

## Tablas con los genes más diferencialmente expresados por condición

Las tablas permiten ordenar y filtrar los genes en función de diversos parámetros. Solamente se incluyen los 100 genes más diferencialmente expresados por cada comparación.

```{r show_annotated_tables, echo=FALSE}

pretty_top_table = function(top_table){
  top_table = dplyr::arrange(top_table, P.Value)
  top_table = dplyr::mutate_if(top_table, is.numeric, signif, digits=2)
  top_table = dplyr::select(top_table, -PROBEID)
}

DT::datatable(pretty_top_table(top_sex_annotated), caption = 'Top Expressed Genes Male vs. Female contrast', filter='top')
DT::datatable(pretty_top_table(top_int_annotated), caption = 'Top Expressed Genes interaction (Sex vs. (Resistant vs. Sensible) contrast', filter='top')
DT::datatable(pretty_top_table(top_res_vs_sens_annotated), caption = 'Top Expressed Genes Resistant vs. Sensible contrast', filter='top')
```

## Volcano plot

Aquí presentamos un volcano plot con algunos extras sobre el original, con los p-valores ajustados y sin ajustar. Además se incluye una tabla con el porcentaje de genes diferencialmente expresados en cada uno de los contrastes.

```{r volcano_plot, echo=FALSE, fig.fullwidth = TRUE, out.width='100%', fig.height=6, out.height='100%', fig.cap='Modified Volcano plot. Horizontal line indicates the limit of significance. Vertical lines indicates the log2 fold difference.'}

# SYMBOLS<- geneSymbols$SYMBOL
# volcanoplot(cm, coef=1, highlight=4, names=SYMBOLS, 
#             main=paste("Differentially expressed genes", colnames(contrast_matrix)[1], sep="\n"))
#   abline(v=c(-1,1))

vol_plt = my_volcano_plot(cm, geneSymbols)

cowplot::plot_grid(plotlist=vol_plt[1:2], ncol=1)

vol_plt$plot_d %>% dplyr::group_by(contrast) %>% dplyr::summarise('Unadjusted p<.05' = paste0(signif(100*mean(sig_p),2),'%'),
                                                          'FDR Corrected p<.05' = paste0(signif(100*mean(sig_adj_p),2),'%')) %>%
  kableExtra::kable(digits = 2, caption='Percentage of significant genes') %>%
  kableExtra::kable_styling(full_width = F) %>% 
  kableExtra::add_footnote('Rounded to two digits')
```


## Diagrama de Venn

```{r venn_diagram, echo=FALSE, fig.fullwidth = TRUE, out.width='100%', fig.height=6, fig.width=6, out.height='100%', fig.cap='Venn Diagram showing how many genes are significant in each of the contrasts and their intersections.'}

res = decideTests(cm)
sum.res.rows<-apply(abs(res),1,sum)
res.selected<-res[sum.res.rows!=0,]
summary(res) %>% kableExtra::kable(caption = 'Differential expression table per contrast') %>% kableExtra::kable_styling(full_width = F)
vennDiagram (res.selected[,1:3], main="Genes in common #1", cex=0.9)
```

## Tabla de anotaciones

```{r annotation_table}
if(!file.exists("anotations.html")){
  library(annaffy)
  genesSelected <- rownames(res.selected)
  at <- aafTableAnn(genesSelected, "drosophila2.db")
  saveHTML (at, "anotations.html","Annotations for selected genes")
  }

```


## Perfiles de expresión

```{r expression_profiles, eval=FALSE, include=FALSE}
library(gplots)

selected_probes = rownames(res.selected)
selected_rows = rownames(exprs(eset_rma_filtered)) %in% selected_probes

HMdata = exprs(eset_rma_filtered) %>%
  tibble::as_tibble(rownames='PROBEID') %>%
  tidylog::inner_join(dplyr::distinct(geneSymbols, PROBEID, .keep_all=T), by=c('PROBEID')) %>% 
  dplyr::filter(PROBEID %in% selected_probes) %>% 
  tibble::column_to_rownames(var='SYMBOL')
HMdata 

  
length(rownames(exprs(eset_rma_filtered)))
length(unique(rownames(exprs(eset_rma_filtered))))




HMdata

geneSymbols

probeNames<-rownames(res)
probeNames.selected<-probeNames[sum.res.rows!=0]
exprs2cluster <-exprs(eset_rma)[probeNames.selected,]

heatmap(exprs2cluster, col=rainbow(100), cexCol=0.9)

```

## Enrichment analsysis

```{r compute_enrichment, echo=FALSE}
library(clusterProfiler)

dros = search_kegg_organism('Drosophila melanogaster', by='scientific_name')

p_tibble =tibble::as_tibble(cm$p.value, rownames='PROBEID') %>% 
  dplyr::mutate(res_vs_sens = p.adjust(res_vs_sens, method = 'fdr'),
                male_vs_female = p.adjust(male_vs_female, method = 'fdr'),
                int_sex_resvssens = p.adjust(int_sex_resvssens, method = 'fdr')) %>% 
  tidylog::inner_join(geneSymbols, by='PROBEID')

gene_list_res_vs_sens = p_tibble %>%
  dplyr::filter(int_sex_resvssens<1) %>% 
  dplyr::pull(ENTREZID) %>%
  bitr(., fromType = "ENTREZID",
        toType = c("ENSEMBL", "SYMBOL"),
        OrgDb = org.Dm.eg.db)
  

gene_list_int = p_tibble %>%
  dplyr::filter(int_sex_resvssens<.05) %>% 
  dplyr::pull(SYMBOL) %>%
  glue::glue('Dmel_{gn}', gn=.)

gene_list_res_vs_sens = p_tibble %>%
  dplyr::filter(res_vs_sens<.1) %>% 
  dplyr::pull(SYMBOL) %>%
  glue::glue('Dmel_{gn}', gn=.)

enrich_sex = gseGO(gene_list_sex, OrgDb = org.Dm.eg.db)
enrich_int = enrichKEGG(gene_list_int, organism='dme', pvalueCutoff = .1)
enrich_res_vs_sens = enrichKEGG(gene_list_res_vs_sens, organism='dme', pvalueCutoff = .1)

data(geneList, package="DOSE")
geneList
gene <- names(geneList)[abs(geneList) > 2]



gene.df <- bitr(p_tibble$ENTREZID, fromType = "ENTREZID",
        toType = c("ENSEMBL", "SYMBOL"),
        OrgDb = org.Dm.eg.db)
geneList

ego2 <- enrichGO(gene         = gene_list_res_vs_sens$ENSEMBL,
                OrgDb         = org.Dm.eg.db,
                keyType       = 'ENSEMBL',
                ont           = "CC",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05)

ego2
```


```{r show_enrichment, echo=FALSE}
goplot(ego2)
  ```


# Introduction

The Tufte handout style is a style that Edward Tufte uses in his books and handouts. Tufte's style is known for its extensive use of sidenotes, tight integration of graphics with text, and well-set typography. This style has been implemented in LaTeX and HTML/CSS^[See Github repositories [tufte-latex](https://github.com/tufte-latex/tufte-latex) and [tufte-css](https://github.com/edwardtufte/tufte-css)], respectively. We have ported both implementations into the [**tufte** package](https://github.com/rstudio/tufte). If you want LaTeX/PDF output, you may use the `tufte_handout` format for handouts, and `tufte_book` for books. For HTML output, use `tufte_html`. These formats can be either specified in the YAML metadata at the beginning of an R Markdown document (see an example below), or passed to the `rmarkdown::render()` function. See @R-rmarkdown for more information about **rmarkdown**.

```yaml
---
title: "An Example Using the Tufte Style"
author: "John Smith"
output:
  tufte::tufte_handout: default
  tufte::tufte_html: default
---
```

There are two goals of this package:

1. To produce both PDF and HTML output with similar styles from the same R Markdown document;
1. To provide simple syntax to write elements of the Tufte style such as side notes and margin figures, e.g. when you want a margin figure, all you need to do is the chunk option `fig.margin = TRUE`, and we will take care of the details for you, so you never need to think about `\begin{marginfigure} \end{marginfigure}` or `<span class="marginfigure"> </span>`; the LaTeX and HTML code under the hood may be complicated, but you never need to learn or write such code.

If you have any feature requests or find bugs in **tufte**, please do not hesitate to file them to https://github.com/rstudio/tufte/issues. For general questions, you may ask them on StackOverflow: http://stackoverflow.com/tags/rmarkdown.

# Headings

This style provides first and second-level headings (that is, `#` and `##`), demonstrated in the next section. You may get unexpected output if you try to use `###` and smaller headings.

`r newthought('In his later books')`^[[Beautiful Evidence](http://www.edwardtufte.com/tufte/books_be)], Tufte starts each section with a bit of vertical space, a non-indented paragraph, and sets the first few words of the sentence in small caps. To accomplish this using this style, call the `newthought()` function in **tufte** in an _inline R expression_ `` `r ` `` as demonstrated at the beginning of this paragraph.^[Note you should not assume **tufte** has been attached to your R session. You should either `library(tufte)` in your R Markdown document before you call `newthought()`, or use `tufte::newthought()`.]

# Figures

## Margin Figures

Images and graphics play an integral role in Tufte's work. To place figures in the margin you can use the **knitr** chunk option `fig.margin = TRUE`. For example:

```{r fig-margin, fig.margin = TRUE, fig.cap = "MPG vs horsepower, colored by transmission.", fig.width=3.5, fig.height=3.5, cache=TRUE, message=FALSE}
library(ggplot2)
mtcars2 <- mtcars
mtcars2$am <- factor(
  mtcars$am, labels = c('automatic', 'manual')
)
ggplot(mtcars2, aes(hp, mpg, color = am)) +
  geom_point() + geom_smooth() +
  theme(legend.position = 'bottom')
```

Note the use of the `fig.cap` chunk option to provide a figure caption. You can adjust the proportions of figures using the `fig.width` and `fig.height` chunk options. These are specified in inches, and will be automatically scaled down to fit within the handout margin.

## Arbitrary Margin Content

In fact, you can include anything in the margin using the **knitr** engine named `marginfigure`. Unlike R code chunks ```` ```{r} ````, you write a chunk starting with ```` ```{marginfigure} ```` instead, then put the content in the chunk. See an example on the right about the first fundamental theorem of calculus.

```{marginfigure}
We know from _the first fundamental theorem of calculus_ that for $x$ in $[a, b]$:
$$\frac{d}{dx}\left( \int_{a}^{x} f(u)\,du\right)=f(x).$$
```

For the sake of portability between LaTeX and HTML, you should keep the margin content as simple as possible (syntax-wise) in the `marginefigure` blocks. You may use simple Markdown syntax like `**bold**` and `_italic_` text, but please refrain from using footnotes, citations, or block-level elements (e.g. blockquotes and lists) there.

Note: if you set `echo = FALSE` in your global chunk options, you will have to add `echo = TRUE` to the chunk to display a margin figure, for example ```` ```{marginfigure, echo = TRUE} ````.

## Full Width Figures

You can arrange for figures to span across the entire page by using the chunk option `fig.fullwidth = TRUE`.

```{r fig-fullwidth, fig.width = 10, fig.height = 2, fig.fullwidth = TRUE, fig.cap = "A full width figure.", warning=FALSE, message=FALSE, cache=TRUE}
ggplot(diamonds, aes(carat, price)) + geom_smooth() +
  facet_grid(~ cut)
```

 Other chunk options related to figures can still be used, such as `fig.width`, `fig.cap`, `out.width`, and so on. For full width figures, usually `fig.width` is large and `fig.height` is small. In the above example, the plot size is $10 \times 2$.

## Main Column Figures

Besides margin and full width figures, you can of course also include figures constrained to the main column. This is the default type of figures in the LaTeX/HTML output.

```{r fig-main, fig.cap = "A figure in the main column.", cache=TRUE}
ggplot(diamonds, aes(cut, price)) + geom_boxplot()
```

# Sidenotes

One of the most prominent and distinctive features of this style is the extensive use of sidenotes. There is a wide margin to provide ample room for sidenotes and small figures. Any use of a footnote will automatically be converted to a sidenote. ^[This is a sidenote that was entered using a footnote.] 

If you'd like to place ancillary information in the margin without the sidenote mark (the superscript number), you can use the `margin_note()` function from **tufte** in an inline R expression. `r margin_note("This is a margin note.  Notice that there is no number preceding the note.")` This function does not process the text with Pandoc, so Markdown syntax will not work here. If you need to write anything in Markdown syntax, please use the `marginfigure` block described previously.

# References

References can be displayed as margin notes for HTML output. For example, we can cite R here [@R-base]. To enable this feature, you must set `link-citations: yes` in the YAML metadata, and the version of `pandoc-citeproc` should be at least 0.7.2. You can always install your own version of Pandoc from http://pandoc.org/installing.html if the version is not sufficient. To check the version of `pandoc-citeproc` in your system, you may run this in R:

```{r eval=FALSE}
system2('pandoc-citeproc', '--version')
```

If your version of `pandoc-citeproc` is too low, or you did not set `link-citations: yes` in YAML, references in the HTML output will be placed at the end of the output document.

# Tables

You can use the `kable()` function from the **knitr** package to format tables that integrate well with the rest of the Tufte handout style. The table captions are placed in the margin like figures in the HTML output.

```{r}
knitr::kable(
  mtcars[1:6, 1:6], caption = 'A subset of mtcars.'
)
```

# Block Quotes

We know from the Markdown syntax that paragraphs that start with `>` are converted to block quotes. If you want to add a right-aligned footer for the quote, you may use the function `quote_footer()` from **tufte** in an inline R expression. Here is an example:

> "If it weren't for my lawyer, I'd still be in prison. It went a lot faster with two people digging."
>
> `r tufte::quote_footer('--- Joe Martin')`

Without using `quote_footer()`, it looks like this (the second line is just a normal paragraph):

> "Great people talk about ideas, average people talk about things, and small people talk about wine."
>
> --- Fran Lebowitz

# Responsiveness

The HTML page is responsive in the sense that when the page width is smaller than 760px, sidenotes and margin notes will be hidden by default. For sidenotes, you can click their numbers (the superscripts) to toggle their visibility. For margin notes, you may click the circled plus signs to toggle visibility.

# More Examples

The rest of this document consists of a few test cases to make sure everything still works well in slightly more complicated scenarios. First we generate two plots in one figure environment with the chunk option `fig.show = 'hold'`:

```{r fig-two-together, fig.cap="Two plots in one figure environment.", fig.show='hold', cache=TRUE, message=FALSE}
p <- ggplot(mtcars2, aes(hp, mpg, color = am)) +
  geom_point()
p
p + geom_smooth()
```

Then two plots in separate figure environments (the code is identical to the previous code chunk, but the chunk option is the default `fig.show = 'asis'` now):

```{r fig-two-separate, ref.label='fig-two-together', fig.cap=sprintf("Two plots in separate figure environments (the %s plot).", c("first", "second")), cache=TRUE, message=FALSE}
```

You may have noticed that the two figures have different captions, and that is because we used a character vector of length 2 for the chunk option `fig.cap` (something like `fig.cap = c('first plot', 'second plot')`).

Next we show multiple plots in margin figures. Similarly, two plots in the same figure environment in the margin:

```{r fig-margin-together, fig.margin=TRUE, fig.show='hold', fig.cap="Two plots in one figure environment in the margin.", fig.width=3.5, fig.height=2.5, cache=TRUE}
p
p + geom_smooth(method = 'lm')
```

Then two plots from the same code chunk placed in different figure environments:

```{r fig-margin-separate, fig.margin=TRUE, fig.cap=sprintf("Two plots in separate figure environments in the margin (the %s plot).", c("first", "second")), fig.width=3.5, fig.height=2.5, cache=TRUE}
knitr::kable(head(iris, 15))
p
knitr::kable(head(iris, 12))
p + geom_smooth(method = 'lm')
knitr::kable(head(iris, 5))
```

We blended some tables in the above code chunk only as _placeholders_ to make sure there is enough vertical space among the margin figures, otherwise they will be stacked tightly together. For a practical document, you should not insert too many margin figures consecutively and make the margin crowded. 

You do not have to assign captions to figures. We show three figures with no captions below in the margin, in the main column, and in full width, respectively.

```{r fig-nocap-margin, fig.margin=TRUE, fig.width=3.5, fig.height=2, cache=TRUE}
# a boxplot of weight vs transmission; this figure
# will be placed in the margin
ggplot(mtcars2, aes(am, wt)) + geom_boxplot() +
  coord_flip()
```
```{r fig-nocap-main, cache=TRUE}
# a figure in the main column
p <- ggplot(mtcars, aes(wt, hp)) + geom_point()
p
```
```{r fig-nocap-fullwidth, fig.fullwidth=TRUE, fig.width=10, fig.height=3, cache=TRUE}
# a fullwidth figure
p + geom_smooth(method = 'lm') + facet_grid(~ gear)
```

# Some Notes on Tufte CSS

There are a few other things in Tufte CSS that we have not mentioned so far. If you prefer `r sans_serif('sans-serif fonts')`, use the function `sans_serif()` in **tufte**. For epigraphs, you may use a pair of underscores to make the paragraph italic in a block quote, e.g.

> _I can win an argument on any topic, against any opponent. People know this, and steer clear of me at parties. Often, as a sign of their great respect, they don't even invite me._
>
> `r quote_footer('--- Dave Barry')`

We hope you will enjoy the simplicity of R Markdown and this R package, and we sincerely thank the authors of the Tufte-CSS and Tufte-LaTeX projects for developing the beautiful CSS and LaTeX classes. Our **tufte** package would not have been possible without their heavy lifting.

You can turn on/off some features of the Tufte style in HTML output. The default features enabled are:

```yaml
output:
  tufte::tufte_html:
    tufte_features: ["fonts", "background", "italics"]
```

If you do not want the page background to be lightyellow, you can remove `background` from `tufte_features`. You can also customize the style of the HTML page via a CSS file. For example, if you do not want the subtitle to be italic, you can define

```css
h3.subtitle em {
  font-style: normal;
}
```

in, say, a CSS file `my_style.css` (under the same directory of your Rmd document), and apply it to your HTML output via the `css` option, e.g.,

```yaml
output:
  tufte::tufte_html:
    tufte_features: ["fonts", "background"]
    css: "my_style.css"
```

There is also a variant of the Tufte style in HTML/CSS named "[Envisoned CSS](http://nogginfuel.com/envisioned-css/)". This style can be used by specifying the argument `tufte_variant = 'envisioned'` in `tufte_html()`^[The actual Envisioned CSS was not used in the **tufte** package. We only changed the fonts, background color, and text color based on the default Tufte style.], e.g.

```yaml
output:
  tufte::tufte_html:
    tufte_variant: "envisioned"
```

To see the R Markdown source of this example document, you may follow [this link to Github](https://github.com/rstudio/tufte/raw/master/inst/rmarkdown/templates/tufte_html/skeleton/skeleton.Rmd), use the wizard in RStudio IDE (`File -> New File -> R Markdown -> From Template`), or open the Rmd file in the package:

```{r eval=FALSE}
file.edit(
  tufte:::template_resources(
    'tufte_html', '..', 'skeleton', 'skeleton.Rmd'
  )
)
```

This document is also available in [Chinese](http://rstudio.github.io/tufte/cn/), and its `envisioned` style can be found [here](http://rstudio.github.io/tufte/envisioned/).

```{r bib, include=FALSE}
# create a bib file for the R packages used in this document
knitr::write_bib(c('base', 'rmarkdown'), file = 'skeleton.bib')
```
